faculty/clients/template.py

```diff
@@ -42,4 +45,310 @@ def publish_new(self, project_id, template, source_directory):
             "sourceDirectory": source_directory,
             "name": template,
         }
-        return self._post_raw(endpoint, json=payload)
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+    def add_to_project_from_directory(
+        self,
+        source_project_id,
+        source_directory,
+        target_project_id,
+        target_directory,
+        parameters,
+    ):
+        endpoint = "project/{}/apply".format(target_project_id)
+        payload = {
+            "sourceProjectId": str(source_project_id),
+            "sourceDirectory": source_directory,
+            "targetDirectory": target_directory,
+            "parameterValues": parameters,
+        }
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+
+def _handle_publishing_response(response):
+    if 200 <= response.status_code < 300:
+        return response
+    elif 400 <= response.status_code < 500:
+        response_body = response.json()
+        error_code = response_body.get("errorCode")
+        if error_code == "resources_validation_failure":
+            raise ResourceValidationFailureResponseSchema().load(response_body)
+        elif error_code == "parameter_validation_failure":
+            raise ParameterValidationFailureSchema().load(response_body)
+        elif error_code == "template_retrieval_failure":
+            raise TemplateRetrievalFailureResponseSchema().load(response_body)
+        elif error_code == "default_parameters_parsing_error":
+            raise DefaultParametersParsingErrorSchema().load(response_body)
+        elif error_code == "generic_parsing_failure":
+            raise GenericParsingErrorSchema().load(response_body)
+        elif error_code == "workspace_files_validation_error":
+            raise WorkspaceFilesValidationErrorSchema().load(response_body)
+    else:
+        raise Exception(
+            "Unexpected response from the server:\n", response.text
+        )
+
+
+class TemplateException(Exception):
+    pass
+
+
+FileTooLargeError = namedtuple(
```

Comment:
In new clients, we're preferring to use `attrs` instead of named tuples for data models. This is probably not worth going back and changing in this PR, but mentioning it for future reference. See `faculty/clients/model.py` for examples. Also FAO @mociarain.

Reply:
Actually, in this particular case this shouldn't be a namedtuple as it seems to be intended to be an exception. In this case, `FileTooLargeError` should inherit from `Exception` to behave properly as an exception.

FWIW, this is easy to achieve with attrs:

```
from attr import attrs, attrib

@attrs
class FileTooLargeError(Exception):
    path = attrib()
    actual_size_bytes = attrib()
    max_bytes = attrib()
```

Reply:
This also applies to `TooManyFilesError` below.

Reply:
Oh, never mind - I see that these are actually nested in the `WorkpaceFilesValidationError` class. Perhaps these named tuples should not have `Error` in the name to avoid suggesting they are themselves exceptions?

Reply:
Yeah, renaming is a good call.